### PR TITLE
tests: ceph_test_rados_api_watch_notify: watch2Delete reconnect

### PR DESCRIPTION
**Pull request by NitzanMordhai**
During test LibRadosWatchNotify.Watch2Delete rados_watch_check can return error -102 if reconnect happened, in that case Broken pipe reconnect and -102 returned

Fixes: https://tracker.ceph.com/issues/51307
Signed-off-by: Nitzan Mordechai <nmordec@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by NitzanMordhai** [comment]
jenkins retest this please

File: src/test/librados/watch_notify.cc [hunk 1]
```diff
@@ -162,7 +162,9 @@ TEST_F(LibRadosWatchNotify, Watch2Delete) {
   }
   ASSERT_TRUE(left > 0);
   ASSERT_EQ(-ENOTCONN, notify_err);
-  ASSERT_EQ(-ENOTCONN, rados_watch_check(ioctx, handle));
+  int rados_watch_check_err = rados_watch_check(ioctx, handle);
+  // We may hit ENOENT due to socket failure and a forced reconnect
+  EXPECT_TRUE(rados_watch_check_err == -ENOTCONN || rados_watch_check_err == -ENOENT);
```
Comment (badone) [inline src/test/librados/watch_notify.cc:167]: Can we add the "  << "Where rados_watch_check_err = " << rados_watch_check_err;" line here so if we do get an unexpected error we can see what it is?

**Comment by badone** [comment]
This looks OK to me. There are other lines where this same change may be applicable but I wonder whether we should add those preemptively or wait and see whether they ever hit this same issue?

**Review by badone — CHANGES_REQUESTED** [review]

**Comment by NitzanMordhai** [comment]
> This looks OK to me. There are other lines where this same change may be applicable but I wonder whether we should add those preemptively or wait and see whether they ever hit this same issue?

i thought about it, I search the tracker and found a few other hits for rados_watch_check, but some of them also returned with 0 when we expected -107, so I think we need another PR for them, are you agree?

File: src/test/librados/watch_notify.cc (hunk 1 above)
Comment (NitzanMordhai) [reply src/test/librados/watch_notify.cc]: done

**Comment by badone** [comment]
> > This looks OK to me. There are other lines where this same change may be applicable but I wonder whether we should add those preemptively or wait and see whether they ever hit this same issue?
> 
> i thought about it, I search the tracker and found a few other hits for rados_watch_check, but some of them also returned with 0 when we expected -107, so I think we need another PR for them, are you agree?

Yes, let's get this in and see what, if anything, pops up next.

**Comment by badone** [comment]
jenkins test docs

**Comment by badone** [comment]
jenkins render docs

**Comment by badone** [comment]
jenkins retest this please

**Review by badone — APPROVED** [review]

**Pull request merged**